File: apps/ios/Tests/QuranRecitationAppTests/MockBackend.swift
import Foundation
@testable import QuranRecitationApp

/// Configurable `QuranBackend` test double. Each method returns the
/// corresponding `Result` field, defaulting to a successful empty
/// payload. Tests override only the methods they exercise.
final class MockBackend: QuranBackend {
  var signedUploadResult: Result<SignedUploadPayload, AppError>?
  var uploadResult: Result<Void, AppError> = .success(())
  var createScoringJobResult: Result<ScoringResultPayload, AppError>?
  var pollScoringResult: Result<ScoringResultPayload, AppError>?
  var surahsResult: Result<[SurahSummary], AppError> = .success([])

  init(
    signedUploadResult: Result<SignedUploadPayload, AppError>? = nil,
    createScoringJobResult: Result<ScoringResultPayload, AppError>? = nil,
    pollScoringResult: Result<ScoringResultPayload, AppError>? = nil,
    surahsResult: Result<[SurahSummary], AppError> = .success([])
  ) {
    self.signedUploadResult = signedUploadResult
    self.createScoringJobResult = createScoringJobResult
    self.pollScoringResult = pollScoringResult
    self.surahsResult = surahsResult
  }

  func requestSignedUploadUrl(filename: String, contentType: String) async throws -> SignedUploadPayload {
    try Self.unwrap(signedUploadResult, fallback: .scoringTimeout)
  }

  func uploadAudio(fileURL: URL, to signedUrl: String) async throws {
    try Self.unwrap(uploadResult, fallback: .scoringTimeout)
  }

  func createScoringJob(uploadKey: String, surahId: String, ayahNumber: Int?) async throws -> ScoringResultPayload {
    try Self.unwrap(createScoringJobResult, fallback: .scoringTimeout)
  }

  func pollScoringResult(jobId: String) async throws -> ScoringResultPayload {
    try Self.unwrap(pollScoringResult, fallback: .scoringTimeout)
  }

  func surahs(limit: Int?, offset: Int?) async throws -> [SurahSummary] {
    switch surahsResult {
    case .success(let value): return value
    case .failure(let error): throw error
    }
  }

  // MARK: - Practice lookups (PR 12)

  /// Closure-based lookup so tests can vary behaviour per id without
  /// stuffing every fixture into a `Result`. Throws fall through to
  /// the caller untouched.
  var surahLookup: ((String) async throws -> SurahSummary?)?
  var ayahLookup: ((String, Int) async throws -> AyahDetail?)?

  func surah(id: String) async throws -> SurahSummary? {
    if let surahLookup { return try await surahLookup(id) }
    return nil
  }

  func ayah(surahId: String, ayahNumber: Int) async throws -> AyahDetail? {
    if let ayahLookup { return try await ayahLookup(surahId, ayahNumber) }
    return nil
  }

  private static func unwrap<T>(_ result: Result<T, AppError>?, fallback: AppError) throws -> T {
    guard let result else { throw fallback }
    switch result {
    case .success(let value): return value
    case .failure(let error): throw error
    }
  }
}

File: apps/android/src/test/java/tv/every/tilawah/android/backend/MockBackend.kt
package tv.every.tilawah.android.backend

import tv.every.tilawah.android.app.AppError
import java.io.File

/**
 * Test double for [QuranBackend]. Each method returns a per-call
 * configurable [Result]; closure-based lookups for surahs / ayahs let
 * tests describe a virtual catalogue without arranging full lists.
 *
 * Mirrors `apps/ios/Tests/QuranRecitationAppTests/MockBackend.swift`.
 */
class MockBackend(
    var surahsResult: Result<List<SurahSummary>> = Result.success(emptyList()),
    var surahLookup: (String) -> Result<SurahSummary?> = { Result.success(null) },
    var ayahLookup: (String, Int) -> Result<AyahDetail?> = { _, _ -> Result.success(null) },
    var ayahsResult: (String) -> Result<List<AyahDetail>> = { Result.success(emptyList()) },
    var referenceAudioResult: (String, Int) -> Result<ReferenceAudio> = { surahId, ayah ->
        Result.failure(AppError.ReferenceUnavailable(surahId, ayah))
    },
    var signedUploadResult: Result<SignedUploadPayload> = Result.failure(
        AppError.BackendUnavailable("requestSignedUploadUrl"),
    ),
    var uploadResult: Result<Unit> = Result.success(Unit),
    var createScoringJobResult: Result<ScoringResultPayload> = Result.failure(
        AppError.BackendUnavailable("createScoringJob"),
    ),
    var fetchScoringJobResult: (String) -> Result<ScoringResultPayload> = {
        Result.failure(AppError.ScoringTimeout)
    },
) : QuranBackend {

    val callLog: MutableList<String> = mutableListOf()

    override suspend fun surahs(limit: Int?, offset: Int?): List<SurahSummary> {
        callLog += "surahs"
        return surahsResult.getOrThrow()
    }

    override suspend fun surah(id: String): SurahSummary? {
        callLog += "surah($id)"
        return surahLookup(id).getOrThrow()
    }

    override suspend fun ayah(surahId: String, ayahNumber: Int): AyahDetail? {
        callLog += "ayah($surahId, $ayahNumber)"
        return ayahLookup(surahId, ayahNumber).getOrThrow()
    }

    override suspend fun ayahs(surahId: String): List<AyahDetail> {
        callLog += "ayahs($surahId)"
        return ayahsResult(surahId).getOrThrow()
    }

    override suspend fun referenceAudio(surahId: String, ayah: Int): ReferenceAudio {
        callLog += "referenceAudio($surahId, $ayah)"
        return referenceAudioResult(surahId, ayah).getOrThrow()
    }

    override suspend fun requestSignedUploadUrl(
        filename: String,
        contentType: String,
    ): SignedUploadPayload {
        callLog += "requestSignedUploadUrl($filename, $contentType)"
        return signedUploadResult.getOrThrow()
    }

    override suspend fun uploadAudio(file: File, signedUrl: String, contentType: String) {
        callLog += "uploadAudio($signedUrl)"
        uploadResult.getOrThrow()
    }

    override suspend fun createScoringJob(
        uploadKey: String,
        surahId: String,
        ayahNumber: Int?,
        sessionId: String?,
    ): ScoringResultPayload {
        callLog += "createScoringJob($surahId, $ayahNumber)"
        return createScoringJobResult.getOrThrow()
    }

    override suspend fun fetchScoringJob(jobId: String): ScoringResultPayload {
        callLog += "fetchScoringJob($jobId)"
        return fetchScoringJobResult(jobId).getOrThrow()
    }
}

/** Convenience builder for a "completed" ScoringResult fixture. */
fun completedScoringResult(
    jobId: String = "job-1",
    score: Double = 0.84,
): ScoringResultPayload = ScoringResultPayload(
    jobId = jobId,
    status = ScoringStatus.Completed,
    score = score,
    verdict = "Mashallah",
    segments = emptyList(),
    feedback = null,
)
